### ISSUE TITLE
Implement docker mirror behaviour

### ISSUE THREAD

**Issue by sergiocharpineljr** [indeedeng-alpha/harbor-container-webhook#7]
Not sure if it is already implemented. It would be nice if the webhook could check if the harbor project is reachable and is working before rewriting container images so it can bypass Harbor in case of a downtime. That's the default behaviour of the Docker mirror funcionality.

**Comment by cnmcavoy** [comment]
I merged your initial implementation in #8 and added some additional settings around the timeout for the http client. If you want to test my changes before I release, that would be helpful.

**Comment by sergiocharpineljr** [comment]
Nice!
I have not tested yet but I think VerifyHarborAPI option in config is not beeing used in code.

**Comment by cnmcavoy** [comment]
Oops, I wired it up and added it to the config examples.

**Comment by sergiocharpineljr** [comment]
I've done some basic tests and found one problem. We are returning an error in verifyHarborIsRunning function when http.client returns error or Harbor API server returns a status code other than 200. That leads to error 500 in the webhook and no pods can be created/updated. What we want is to bypass Harbor is these situations.

So I think we should remove the error in verifyHarborIsRunning.

```
diff --git a/internal/static/transfomer.go b/internal/static/transfomer.go
index 318ee9c..40c87b0 100644
--- a/internal/static/transfomer.go
+++ b/internal/static/transfomer.go
@@ -2,7 +2,6 @@ package static

 import (
        "crypto/tls"
-       "errors"
        "net/http"

        "github.com/hashicorp/go-cleanhttp"
@@ -14,25 +13,25 @@ import (
 type staticTransformer struct {
        proxyMap       map[string]string
        harborEndpoint string
-       HarborVerifier func(string) (bool, error)
+       HarborVerifier func(string) (bool)
 }

 type harborCheck struct {
        client *http.Client
 }

-func (h *harborCheck) verifyHarborIsRunning(endpoint string) (bool, error) {
+func (h *harborCheck) verifyHarborIsRunning(endpoint string) (bool) {
        res, err := h.client.Get(endpoint + "/api/version")
        if res != nil && res.Body != nil {
                defer res.Body.Close()
        }
        if err != nil {
-               return false, err
+               return false
        }
        if res.StatusCode != 200 {
-               return false, errors.New("harbor API server did not return 200 status code")
+               return false
        }
-       return true, nil
+       return true
 }

 func (s *staticTransformer) RewriteImage(imageRef string) (string, error) {
@@ -41,7 +40,7 @@ func (s *staticTransformer) RewriteImage(imageRef string) (string, error) {
                return "", err
        }

-       running, err := s.HarborVerifier(s.harborEndpoint)
+       running := s.HarborVerifier(s.harborEndpoint)
        if running {
                if rewrite, ok := s.proxyMap[registry]; ok {
                        return webhook.ReplaceRegistryInImageRef(imageRef, rewrite)
@@ -69,8 +68,8 @@ func NewTransformer(conf config.StaticProxy) webhook.ContainerTransformer {
        harborCheck := &harborCheck{client: client}
        harborVerifier := harborCheck.verifyHarborIsRunning
        if !conf.VerifyHarborAPI {
-               harborVerifier = func(string) (bool, error) {
-                       return true, nil
+               harborVerifier = func(string) (bool) {
+                       return true
                }
        }

diff --git a/internal/static/transfomer_test.go b/internal/static/transfomer_test.go
index ac0ccdc..14995c0 100644
--- a/internal/static/transfomer_test.go
+++ b/internal/static/transfomer_test.go
@@ -13,15 +13,15 @@ func TestStaticTransformer_Ready(t *testing.T) {
        require.Nil(t, transformer.Ready())
 }

-func MockHarborVerifierTrue() func(string) (bool, error) {
-       return func(string) (bool, error) {
-               return true, nil
+func MockHarborVerifierTrue() func(string) (bool) {
+       return func(string) (bool) {
+               return true
        }
 }

-func MockHarborVerifierFalse() func(string) (bool, error) {
-       return func(string) (bool, error) {
-               return false, nil
+func MockHarborVerifierFalse() func(string) (bool) {
+       return func(string) (bool) {
+               return false
        }
 }
```

**Comment by cnmcavoy** [comment]
Good point on the error handling. I agree that we should handle the 500/400 http code cases better. 

My initial thought is to leave the error handling in the harbor check in place, and use the error for printing a warning to the logs (for debugging) and but swallow the error in `RewriteImage` transformer. I'll push a commit that does that and we can see.

**Comment by sergiocharpineljr** [comment]
I've repeated the basics tests and it is working now. Thanks! I'll be waiting for the next release.

Although I'm only using static mode it would be good to have the same behaviour on dynamic mode. I haven't included it in my PR because I couldn't get it working in my env and didn't have time to debug further.

**Comment by brettmorien** [comment]
This looks very useful. Could you cut a release and a new docker image?

**Comment by brettmorien** [comment]
This looks very useful. Could you please cut a release and a new docker image?

**Comment by cnmcavoy** [comment]
@sergiocharpineljr you might be interested in the work I am doing in https://github.com/indeedeng-alpha/harbor-container-webhook/tree/cnmcavoy/rules-rewrite

I decided after evaluating the webhook for the last year, that the dynamic mode was not really what we wanted. The static mode has proven very useful, but its limited in some ways (we started wanting to be able to use regular expressions) and it made more sense to rewrite the core logic for that.

The new regex rules based engine has support for checking if the image exists when the container ref is being rewritten (which seems like the intent of this issue, unless I'm mistaken) and otherwise is fairly similar to the previous static mode.

I'll probably tag and publish the new engine after evaluating it a bit further in a week or two.
